### DISCUSSION TITLE
Passing nlohmann object array item to a function

### DISCUSSION THREAD

**Discussion by Shielsy** [nlohmann/json#3302] [Q&A] [answered]
I'm want to pass each item in a json object to a function so I can parse each item. The only problem is I don't know what type the functions should accept. See the example below. What parameter should function `parse` accept? Thanks in advance.

```cpp

#include "nlohmann/json.hpp"
#include <iostream>

namespace
{
const auto testJson = R"(
     {
        "test":
        {
           "1":
           {
              "name" : "test object 1"
           },
           "2":
           {
              "name" : "test object 2"
           }
        }
    })"_json;
}

int main(int argc, char** argv)
{
    const auto parse = [](const nlohmann::json &item) //    << --- what type should this parameter be?
    {
         std::cout << item.key() << std::endl; 
    };

    for (const auto &item : testJson.at("test").items())
    {
        parse(item);   /// compilation error
        std::cout << item.key() << std::endl; // this compiles OK
    }

    return 0;
}
```

**Comment by gregmarr** [comment]
The actual type is `nlohmann::detail::iteration_proxy_value<nlohmann::json::iterator>` or `nlohmann::detail::iteration_proxy_value<nlohmann::json::const_iterator>` depending on whether or not testJson is `const`, but you REALLY don't want to be writing that, especially since it's digging into the `detail` namespace.  Your best bet is generic lambdas or a regular template class/function:

    const auto parse = [](const auto &item)
    {
         std::cout << item.key() << std::endl; 
    };

or the two pieces separately:

    const auto parse = [](const std::string &key, const nlohmann::json &value)
    {
         std::cout << key() << std::endl; 
    };

        parse(item.key(), item.value());

**Reply by nlohmann** [reply]
`decltype` can also help.

**Comment by Shielsy** [comment, marked as answer]
Thanks for the reply.

I will probably go with 'parse(item.key(), item.value());'
